vm.go

```diff
@@ -617,17 +618,24 @@ func (vm *vm) run() {
 				ctx = context.Background()
 			}
 
-			if waitErr := vm.r.limiter.WaitN(ctx, vm.r.limiterTicksLeft); waitErr != nil {
-				if vm.r.vm.ctx == nil {
-					panic(waitErr)
-				}
-				if ctxErr := vm.r.vm.ctx.Err(); ctxErr != nil {
-					panic(ctxErr)
-				}
-				if strings.Contains(waitErr.Error(), "would exceed") {
-					panic(context.DeadlineExceeded)
+			now := time.Now()
+			r := vm.r.limiter.ReserveN(now, vm.r.limiterTicksLeft)
+			if !r.OK() {
+				panic("failed to make reservation")
+			}
+
+			if delay := r.DelayFrom(now); delay > 0 {
+				vm.r.limiterWaitCount++
+				vm.r.limiterWaitTotalTime += delay
+
+				if err := sleep(ctx, delay); err != nil {
+					r.Cancel()
+					panic(err)
 				}
-				panic(waitErr)
+			}
+			select {
+			case <-ctx.Done():
+				panic(ctx.Err())
```

Comment:
[opt] I don't believe this extra check is necessary if `delay == 0`. We already interrupt the VM [here](https://github.com/10gen/baas/blob/a2b4a5adb7a3d713683b801fdbd73818d7b84c15/function/execution/vm/context.go#L429), when the context times out / is canceled.

Reply:
Hmm this was added in response to [this test](https://github.com/10gen/baas/blob/ec4c390015aa62ac658ad48547e70a0cd3e241f4/function/execution/vm/execution_test.go#L1994) failing.. Though I still have test failures so I need to investigate what is happening.

Reply:
https://spruce.mongodb.com/version/65df6c7e0ae606b392c17522/task-duration?duration=DESC

Reply:
Sorry, I'm not sure I follow. Did you figure out what caused the test to fail? I still think we can get rid of this extra select statement. Correct me if I'm wrong.

Reply:
Ok tests are finally [passing](https://spruce.mongodb.com/version/65df7a90a4cf47015154653f/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC) with these changes, test-api-other and test-vm-goja tasks do not finish without the extra context check ([evergreen](https://spruce.mongodb.com/version/65df6c7e0ae606b392c17522/task-duration?duration=DESC)).

This check also exists in the WaitN function, so we're just pulling that over here.
![image](https://github.com/mongodb-forks/goja/assets/6693710/06dd0e09-952b-4826-858b-e6479da2af13)

Reply:
Let me know if you'd like to chat about this.

Reply:
Specifically the [TestFunctinExecTimeLimit](https://github.com/10gen/baas/blob/ec4c390015aa62ac658ad48547e70a0cd3e241f4/function/execution/vm/execution_test.go#L1983) test is what never finishes.

Reply:
I see. [opt] The `wait` function has the select statement before it performs the reserve/wait operations. Should we do the same here? i.e. have this select statement right after we define `ctx`.

Reply:
Yeah I think that's a good idea to stay consistent.